mewarpx/mewarpx/mwxrun.py

```diff
@@ -56,6 +58,51 @@ def init_grid(self, xmin, xmax, zmin, zmax, nx, nz, max_grid_size=4):
         self._set_geom_str()
         self._set_grid_params()
 
+    def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
+        """Calculate timestep size based on grid data and CFL parameter
+
+         Arguments:
+            V_anode (float): Vacuum bias of anode relative to cathode, V.
+            DT (float): The dt of each step of the simulation, if not given it
+                will be calculated.
+            CFL_factor (float): Multiplier to determine the actual timestep
+                given the CFL ratio. eg. ``dt = CFL_factor * CFL_ratio``.
+            V_grid (float): Vaccum bias of highest-voltage grid relative to
+                cathode, V. If not defined a value of 5 which is a safe value
+                for finding vmax in the case that the electron velocities are
+                only set by their temperature
+        """
+        if self.grid is None:
+            raise ValueError("init_grid must be called before init_timestep!")
+
+        if DT is not None:
+            self.simulation.time_step_size = DT
+            return DT
+
+        if CFL_factor is None:
+            raise ValueError("No CFL_factor passed into init_timestep when DT is not defined!")
+
+        vmax = np.sqrt(
+               2*np.e/constants.m_e * max(0.0, V_anode, V_grid)
+        )
+
+        if self.grid.nx > 0:
+            dx = (self.grid.xmax - self.grid.xmin) / self.grid.nx
+        else:
+            dx = 0.
+        # The picmi standard uses x and y for 2d, while we use x and z
+        dz = (self.grid.ymax - self.grid.ymin) / self.grid.ny
+
+        if dx > 0.:
+            dt_local = min(dx, dz) / vmax * CFL_factor
+        else:
+            dt_local = dz / vmax * CFL_factor
```

Comment:
Since this function has to be run after `init_grid()` (which also runs `_set_grid_params()`) you can just use `self.dx` instead of recalculating dx, and same for dz. Also, WarpX doesn't have functionality for 1D runs so you don't need the `if dx <= 0` part of the call below. (I didn't test this suggestion so please confirm that it works)
```suggestion
        if self.geom_str == 'XZ' or self.geom_str == 'RZ' :
            dt_local = min(self.dx, self.dz) / vmax * CFL_factor
        if self.geom_str == 'XYZ':
            dt_local = min(self.dx, self.dy, self.dz) / vmax * CFL_factor
```